crates/astria-sequencer/src/app/vote_extension.rs

```diff
@@ -0,0 +1,649 @@
+                tracing::error!(error = %e, "failed to verify vote extension");
+                abci::response::VerifyVoteExtension::Reject
+            }
+        }
+    }
+}
+
+// see https://github.com/skip-mev/slinky/blob/5b07f91d6c0110e617efda3f298f147a31da0f25/abci/ve/utils.go#L24
+async fn verify_vote_extension<S: StateReadExt>(
+    state: &S,
+    oracle_vote_extension_bytes: bytes::Bytes,
+    is_proposal_phase: bool,
+) -> anyhow::Result<()> {
+    let oracle_vote_extension = RawOracleVoteExtension::decode(oracle_vote_extension_bytes)
+        .context("failed to decode oracle vote extension")?;
+    let max_num_currency_pairs =
+        DefaultCurrencyPairStrategy::get_max_num_currency_pairs(state, is_proposal_phase)
+            .await
+            .context("failed to get max number of currency pairs")?;
+
+    ensure!(
+        oracle_vote_extension.prices.len() as u64 <= max_num_currency_pairs,
+        "number of oracle vote extension prices exceeds max expected number of currency pairs"
+    );
+
+    for prices in oracle_vote_extension.prices.values() {
+        ensure!(
+            prices.len() <= MAXIMUM_PRICE_BYTE_LEN,
+            "encoded price length exceeded {MAXIMUM_PRICE_BYTE_LEN}"
+        );
+    }
+
+    Ok(())
+}
+
+// see https://github.com/skip-mev/slinky/blob/158cde8a4b774ac4eec5c6d1a2c16de6a8c6abb5/abci/ve/vote_extension.go#L290
+async fn transform_oracle_service_prices<S: StateReadExt>(
+    state: &S,
+    prices: QueryPricesResponse,
+) -> anyhow::Result<OracleVoteExtension> {
+    let mut strategy_prices = IndexMap::new();
+    for (currency_pair_id, price_string) in prices.prices {
+        let currency_pair = currency_pair_id
+            .parse()
+            .context("failed to parse currency pair")?;
+
+        // prices are encoded as just a decimal string in the sidecar response
+        let price: u128 = price_string
+            .parse()
+            .context("failed to parse price string")?;
+
+        let Ok(id) = DefaultCurrencyPairStrategy::id(state, &currency_pair).await else {
+            trace!(
+                currency_pair = currency_pair.to_string(),
+                "currency pair not found in state; skipping"
+            );
+            continue;
+        };
+        let encoded_price = DefaultCurrencyPairStrategy::get_encoded_price(state, price);
+
+        debug!(
+            currency_pair = currency_pair.to_string(),
+            id, price, "transformed price for inclusion in vote extension"
+        );
+        strategy_prices.insert(id, encoded_price.into());
+    }
+
+    Ok(OracleVoteExtension {
+        prices: strategy_prices,
+    })
+}
+
+pub(crate) struct ProposalHandler;
+
+impl ProposalHandler {
+    // called during prepare_proposal
+    pub(crate) async fn prune_and_validate_extended_commit_info<S: StateReadExt>(
+        state: &S,
+        height: u64,
+        mut extended_commit_info: ExtendedCommitInfo,
+    ) -> anyhow::Result<ExtendedCommitInfo> {
```

Comment:
To make the types clearer: this should return a `ValidatedExtendedCommitInfo`. It can just be a thin wrapper `struct ValidatedExtendedCommitInfo(ExtendedCommitInfo).`